bin/ubuntu-core-initramfs

```diff
@@ -614,8 +615,54 @@ def create_initrd_pkg_list(dest_dir, sysroot):
         pkg_list.write(out)
 
 
+def mkcpio(path, output, *, compress=True):
+    cpio_cmd = subprocess.Popen(['cpio', '--null', '--reproducible', '--create', '--quiet', '--format=newc', '--owner=0:0'],
+                                stdin=subprocess.PIPE, stdout=subprocess.PIPE if compress else output,
+                                cwd=path)
+
+    if compress:
+        zstd_cmd = subprocess.Popen(['zstd', '-1', '-T0'],
+                                    stdin=cpio_cmd.stdout, stdout=output)
+    else:
+        zstd_cmd = None
+
+    file_list = []
+    for root, dirs, files in os.walk(path):
+        for d in dirs:
+            file_list.append(os.path.join(os.path.relpath(root, path), d))
+        for f in files:
+            file_list.append(os.path.join(os.path.relpath(root, path), f))
+    file_list = sorted(file_list)
+    for path in file_list:
+        cpio_cmd.stdin.write(path.encode('ascii'))
+        cpio_cmd.stdin.write(b'\0')
+
+    cpio_cmd.stdin.close()
+    if zstd_cmd is not None:
+        zstd_cmd.wait()
+    cpio_cmd.wait()
+    if cpio_cmd.returncode != 0:
+        raise RuntimeError("cpio failed")
+    if zstd_cmd is not None:
+        if zstd_cmd.returncode != 0:
+            raise RuntimeError("zstd failed")
+
+def mk_amd_ucode(output):
+    for path in glob.glob('/lib/firmware/amd-ucode/*.bin'):
+        with open(path, 'rb') as f:
+            shutil.copyfileobj(f, output)
+
+def mk_intel_ucode(output_path):
+    check_call(['iucode_tool', f'--write-to={output_path}', '/lib/firmware/intel-ucode/'])
```

Comment:
It is weird to use `--write-earlyfw` since we want to combine the AMD firmware with the Intel firmware in the same early initrd. `--write-earlyfw` would create the initrd only with Intel ones.

Reply:
I have the impression that initramfs-tools chooses what firmware to use dependending on the processor of the current machine. It is a different script for AMD: `/usr/share/initramfs-tools/hooks/amd64_microcode`

Reply:
Also, those are hooks for initramfs-tools. They are not expected to be run as tools. We would need to provide environment variables to mimic the initramfs-tools API. And I feel it can break easily.